File: server/modules/subscription/monitor.ts
import { checkPaymentStatus } from "./nowpayments.service";
import { SubscriptionService } from "./subscription.service";
import { AccountService } from "../account/account.service";

const SCAN_INTERVAL_MS = 60_000; // 1 minute

/**
 * Start the monitor that periodically checks pending payments with NowPayments API.
 */
export function startMonitor() {
    console.log("[Monitor] Starting payment monitor...");

    // Run immediately, then on interval
    runOnce();
    setInterval(runOnce, SCAN_INTERVAL_MS);
}

async function runOnce() {
    try {
        await checkPendingPayments();
    } catch (err) {
        console.error("[Monitor] Error:", err);
    }
}

/**
 * Poll NowPayments API for pending payments that have a real payment_id.
 * When payment is confirmed, mark as confirmed.
 * When expired, mark as expired.
 */
async function checkPendingPayments() {
    const pendingRecords = await SubscriptionService.findPendingRecords();
    for (const record of pendingRecords) {
        try {
            if (!record.payment_id) {
                if (Date.now() - record.create_time > 30 * 60 * 1000) {
                    await SubscriptionService.updateRecordByTxid(record.txid, { status: "expired" });
                    console.log(new Date(), `[Monitor] Payment expired for ${record.account_id}`);
                    continue;
                }
                continue;
            }
            const { status } = await checkPaymentStatus(record.payment_id);
            if (status === "confirmed") {
                // Balance is computed from transaction records — no manual update needed

                await SubscriptionService.updateRecordByTxid(record.txid, {
                    status: "confirmed",
                    confirmations: 1,
                });
                console.log(`[Monitor] Payment confirmed for ${record.account_id}`);
            } else if (status === "expired") {
                await SubscriptionService.updateRecordByTxid(record.txid, { status: "expired" });
                console.log(`[Monitor] Payment expired for ${record.account_id}`);
            }
        } catch (err) {
            console.log(`[Monitor] Check payment ${record.payment_id} failed:`, (err as Error).message);
        }
    }
}

File: server-go/internal/monitor/monitor.go
// Payment monitor — polls NowPayments for pending invoices every minute
// (subscription/monitor.ts).
package monitor

import (
	"database/sql"
	"fmt"
	"time"

	"onekey/server/internal/service"
	"onekey/server/internal/store"
)

func Start(db *sql.DB, settings *service.Settings) {
	fmt.Println("[Monitor] Starting payment monitor...")
	go func() {
		runOnce(db, settings)
		ticker := time.NewTicker(60 * time.Second)
		defer ticker.Stop()
		for range ticker.C {
			runOnce(db, settings)
		}
	}()
}

func runOnce(db *sql.DB, settings *service.Settings) {
	if err := checkPendingPayments(db, settings); err != nil {
		fmt.Println("[Monitor] Error:", err)
	}
}

func checkPendingPayments(db *sql.DB, settings *service.Settings) error {
	pending, err := store.TxPending(db)
	if err != nil {
		return err
	}
	for _, record := range pending {
		if record.PaymentID == "" {
			if store.Now()-record.CreateTime > 30*60*1000 {
				_ = store.TxUpdateByTxid(db, record.Txid, map[string]any{"status": "expired"})
				fmt.Println(time.Now(), "[Monitor] Payment expired for", record.AccountID)
			}
			continue
		}
		status, _, err := service.NowPaymentsStatus(settings, record.PaymentID)
		if err != nil {
			fmt.Println("[Monitor] Check payment", record.PaymentID, "failed:", err)
			continue
		}
		switch status {
		case "confirmed":
			_ = store.TxUpdateByTxid(db, record.Txid, map[string]any{
				"status": "confirmed", "confirmations": 1,
			})
			if err := store.AccountAddBalance(db, record.AccountID, record.Amount); err != nil {
				fmt.Println("[Monitor] balance update failed:", err)
				continue
			}
			fmt.Println("[Monitor] Payment confirmed for", record.AccountID)
		case "expired":
			_ = store.TxUpdateByTxid(db, record.Txid, map[string]any{"status": "expired"})
			fmt.Println("[Monitor] Payment expired for", record.AccountID)
		}
	}
	return nil
}
